server/error_recorder.go

```diff
@@ -0,0 +1,49 @@
+package server
+
+import (
+	"net/http"
+	"runtime/debug"
+	"time"
+
+	"github.com/bytebase/bytebase/api"
+
+	"github.com/labstack/echo/v4"
+)
+
+func errorRecorderMiddleware(err error, s *Server, c echo.Context, e *echo.Echo) {
+	e.DefaultHTTPErrorHandler(err, c)
+
+	req := c.Request()
+	res := c.Response()
+
+	he, isHTTPError := err.(*echo.HTTPError)
+
+	if res.Status == http.StatusInternalServerError {
+		var role api.Role
+		if r, ok := c.Get(getRoleContextKey()).(api.Role); ok {
+			role = r
+		}
+
+		var stackTrace string
+		// There're basically two kinds of errors: internal HTTP errors and runtime panics.
+		// If we encounter runtime panics, we can report the stack trace for debugging.
+		if !isHTTPError || he.Internal == nil {
+			// Deal with runtime panics.
+			// We get a snapshot of the stack trace here, which will include where the panic occurred.
+			stackTrace = string(debug.Stack())
```

Comment:
Why don't we use string(debug.Stack()) on line 45?

Reply:
The stack trace is only helpful to us when a panic occurs. If it is an internal HTTP error, we can find the problem by its error message.

Reply:
But I doesn't hurt to keep the stack. Or in other words, we are just setting an empty string when the stack is zero bytes.

Reply:
When we use `string(debug.Stack())`, we can always get a long string of stack trace. This is not an empty string of zero bytes. It won't indicate where the internal error is from, either. Do we still need it?

Reply:
I see what you mean. Let's update the comment above indicating the stack is only meaningful for panics.